File: answers/leetcode/Remove Nth Node From End of List/Remove Nth Node From End of List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode dummy = new ListNode(-1);
        dummy.next = head;
        ListNode right = dummy;
        for (int i = 0; i <= n && right != null; i++) {
            right = right.next;
        }
        ListNode left = dummy;
        while (right != null) {
            left = left.next;
            right = right.next;
        }
        if (left != null && left.next != null) {
            left.next = left.next.next;
        }
        return dummy.next;
    }
}

File: answers/leetcode/Remove Nth Node From End of List/Remove Nth Node From End of List.cpp
//@result 207 / 207 test cases passed. Status: Accepted Runtime: 4 ms Submitted: 0 minutes ago You are here! Your runtime beats 24.78% of cpp submissions.

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *removeNthFromEnd(ListNode *head, int n) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
        ListNode *fast = head, *slow = head;
        for (int i = 0; i < n; ++ i)
        {
			fast = fast->next;
		}
		if (NULL == fast)
		{
			return head->next;
		}
		while (fast->next)
		{
			fast = fast->next;
			slow = slow->next;
		}
		slow->next = slow->next->next;
		return head;
    }
};
